[darrentfy] iP

darrentfy opened the pull request:
## DukePro
> "And perhaps, what is this?" - DJ Khaled

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~fast~~ **SUPER FAST** to use

All you need to do is,
1. download it from [here](https://github.com/darrentfy/ip/releases/tag/v0.1)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :on::top::speaking_head::speaking_head::fire::fire::fire::100::100::100::bangbang::bangbang:

And it is **FREE!**

Features:
- [x] manage tasks
- [x] search your tasks
- [x] mark and unmark your tasks

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public static void main(String[] args) {
    new Duke("./data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
renamed chatbot and implemented skeletal version

Commit:
added feature to echo commands

Commit:
added feature to add and display tasks

Commit:
added Task class to represent tasks and feature to mark and unmark tasks as done

Commit:
added subclasses of the Task class, Deadline, Event and Todo

Commit:
added custom exception classes and handling of incorrect inputs

Commit:
added feature to delete tasks

Commit:
reorganised code and and removed exceptions that could be handled with if else statements

Commit:
added DukeException class to handle exceptions

Commit:
no message

Commit:
Merge branch 'master' of github.com:darrentfy/ip

# Conflicts:
#	src/main/java/Duke.java

Commit:
Added DukeException class to handle exceptions

Commit:
added text ui testing

Commit:
Add checks for whitespace after command in user input eg "todo "

Commit:
Add feature to save tasks list

Commit:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add feature to save tasks list

Commit:
Add feature to read date and time and update testcases

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add feature to read date and time and update testcases

Commit:
Create new classes to refactor code

Commit:
Merge branch 'branch-moreOOP'

Commit:
Organise classes into task, command and duke packages

Commit:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552'

Commit:
Set up Gradle and add unit testing for Parser and TaskList

Commit:
Add JavaDoc comments

Commit:
Reorganised code to fit SE-EDU Java coding standard

Commit:
Add feature to filter tasks using find keyword

Commit:
Merge branch 'branch-A-Level-9'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-Javadoc'

Commit:
Add checkstyle and fixed coding style violations

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add GUI to Duke using JavaFX

Commit:
Merge branch 'branch-Level-10'

jonyeokj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,179 @@
+ * Parses user input
+ */
+public class Parser {
+    private static final DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
```
Good use of a constant to avoid magic strings/numbers in code to enhance readability and maintainability.

jonyeokj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,179 @@
+            case FindCommand.COMMAND_WORD:
+                return prepareFind(details);
+            default:
+                throw new DukeException("Sorry! I do not recognise this command");
```
Throwing a DukeException makes it clear that it is specific to this context, good and comprehensive error handling. 👍

jonyeokj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,179 @@
+     */
+    private static MarkCommand prepareMark(String[] details) throws DukeException {
+        // user input only has the command eg "mark"
+        if (details.length < 2 || details[1].trim().isEmpty()) {
```
The code has repeated checks for details length and emptiness. Consider creating a helper method to standardize these checks.
```java
private static void validateInputDetails(String[] details) throws DukeException {
    if (details.length < 2 || details[1].trim().isEmpty()) {
        throw new DukeException("Invalid command! Please provide more details");
    }
}
```

jonyeokj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,179 @@
+            throw new DukeException("Invalid command! "
+                    + "Please include the index of the task you wish to mark");
+        }
+        int markTaskId = Integer.parseInt(details[1]) - 1;
```
Consider adding checks for Integer.parseInt. If the user enters a non-integer, this would throw an unhandled exception.

jonyeokj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,109 @@
+     *
+     * @throws DukeException when data of invalid format is read from Storage file
+     */
+    public ArrayList<Task> readFile() {
```
You might want to further break down some of the logic in readFile into smaller methods, to promote modularity and readability. Just a nit.

jonyeokj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,109 @@
+                case "T":
+                    task = new Todo(description);
+                    break;
+                case "D":
```
Might be better to encapsulate readFile for each different type of task in their own file. It will be cleaner if each task had a static factory method for creation from a string.

jonyeokj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,109 @@
+                }
+                tasks.add(task);
+            }
+        } catch (IOException | DukeException | DateTimeParseException e) {
```
Error handling for DateTimeParseException is generic. It might be beneficial to specify that the date format in the file is wrong.

jonyeokj reviewed (commented):
Code is well-documented and has clear javadoc descriptions. Areas of improvement I would suggest are nits. I suggest better error handling at certain parts of the code, as well as encapsulation in the task readFile method.

Overall, code LGTM! 👍

LINCHENYU2030S commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+package command;
```
I like how you organise all the classes related to command into one folder.

LINCHENYU2030S commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,49 @@
+ * Deletes a task from the todo list
+ */
+public class DeleteCommand extends Command {
+    public static final String COMMAND_WORD = "delete";
+    public static final String MESSAGE_SUCCESS = "Noted. I've removed this task:\n";
```
I like how the constant names are all uppercased.

LINCHENYU2030S commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+        return " ";
+    }
+
+    public String getDescription() {
+        return this.description;
+    }
```
i like how the method is named consistently with the field description

LINCHENYU2030S commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+        return this.description;
+    }
+
+    public String getDetails() {
+        return this.description;
+    }
```
not sure why we need this method because this method does the same thing as the method getDescription implemented right above it

LINCHENYU2030S commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+    protected String description;
+
+    /** True if the task is done, false otherwise */
+    protected boolean isDone;
```
I like how the variable names are named in camelCase

LINCHENYU2030S reviewed (commented):
LGTM. I like how OOP is well-applied in this change. I like how the code is organized into several packages.

Commit:
Refactor task classes to remove an unnecessary method

Commit:
Refactor Task classes and add Java assert

Java assert feature is not used and task classes have a getType method
to return a string with a letter representing the task type.

While there are checks for invalid user input values, there is no way to
check for any invalid program states. The getType method is repetitive
as it simply returns a single letter depending on the Task subclass.

Let's,
* add Java assertions to monitor program state
* refactor Task subclasses to have a taskType field instead

Adding assertions allow us to detect possible bugs and refactoring the
Task subclasses helps reduce code duplication from unnecessary methods.

Commit:
Merge branch 'branch-A-Assertions'

Commit:
Refactor code to improve code quality

Parser has multiple repeated checks and long conditional checks.

These unnecessarily repetitive and complicated checks make code less
readable.

Let's abstract the checks into a method and rewrite repetitive messages
as constants.

These changes will make the code more readable and easier to understand.

darrentfy committed:
Merge pull request #1 from darrentfy/branch-A-CodeQuality

Refactor code to improve code quality

Commit:
Merge branch 'master' of github.com:darrentfy/ip

Commit:
Set up CI with GitHub Actions

Commit:
Remove part of CI related to I/O redirection

Commit:
Add user help feature

Commit:
Merge branch 'branch-C-Help'

Commit:
Fix issues with checkstyle

Commit:
Refactor Help command for individual commands

Commit:
Add User Guide
